// @sanity/desk-tool/lib/panes/documentPane/statusBar/documentStatusBarActions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DocumentStatusBarActions = DocumentStatusBarActions;
exports.HistoryStatusBarActions = HistoryStatusBarActions;

var _react = _interopRequireWildcard(require("react"));

var _reactHooks = require("@sanity/react-hooks");

var _ui = require("@sanity/ui");

var _default = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _hotkeys = _interopRequireDefault(require("part:@sanity/components/typography/hotkeys"));

var _utils = require("part:@sanity/base/actions/utils");

var _resolver = _interopRequireDefault(require("part:@sanity/base/document-actions/resolver"));

var _HistoryRestoreAction = require("../../../actions/HistoryRestoreAction");

var _actionMenu = require("./actionMenu");

var _actionStateDialog = require("./actionStateDialog");

var _documentStatusBarActions = _interopRequireDefault(require("./documentStatusBarActions.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _toArray(arr) { return _arrayWithHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && iter[Symbol.iterator] != null || iter["@@iterator"] != null) return Array.from(iter); }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

// eslint-disable-next-line complexity
function DocumentStatusBarActionsInner(props) {
  var states = props.states,
      showMenu = props.showMenu;

  var _states = _toArray(states),
      firstActionState = _states[0],
      menuActionStates = _states.slice(1);

  var _useState = (0, _react.useState)(null),
      _useState2 = _slicedToArray(_useState, 2),
      buttonContainerElement = _useState2[0],
      setButtonContainerElement = _useState2[1];

  return /*#__PURE__*/_react.default.createElement("div", {
    className: props.isMenuOpen ? _documentStatusBarActions.default.isMenuOpen : _documentStatusBarActions.default.root
  }, firstActionState && /*#__PURE__*/_react.default.createElement("div", {
    className: _documentStatusBarActions.default.mainAction
  }, /*#__PURE__*/_react.default.createElement(_ui.Tooltip, {
    disabled: !(firstActionState.title || firstActionState.shortcut),
    content: /*#__PURE__*/_react.default.createElement(_ui.Flex, {
      padding: 2,
      style: {
        maxWidth: 300
      }
    }, firstActionState.title, firstActionState.shortcut && /*#__PURE__*/_react.default.createElement(_ui.Box, {
      marginLeft: 2
    }, /*#__PURE__*/_react.default.createElement(_hotkeys.default, {
      keys: String(firstActionState.shortcut).split('+'),
      size: "small"
    }))),
    portal: true,
    placement: "top"
  }, /*#__PURE__*/_react.default.createElement("div", {
    ref: setButtonContainerElement
  }, /*#__PURE__*/_react.default.createElement(_default.default, {
    className: showMenu ? _documentStatusBarActions.default.mainActionButtonWithMoreActions : _documentStatusBarActions.default.mainActionButton,
    icon: firstActionState.icon,
    color: firstActionState.disabled ? undefined : firstActionState.color || 'success',
    disabled: props.disabled || Boolean(firstActionState.disabled),
    "aria-label": firstActionState.title,
    onClick: firstActionState.onHandle
  }, firstActionState.label))), firstActionState.dialog && /*#__PURE__*/_react.default.createElement(_actionStateDialog.ActionStateDialog, {
    dialog: firstActionState.dialog,
    referenceElement: buttonContainerElement
  })), showMenu && menuActionStates.length > 0 && /*#__PURE__*/_react.default.createElement(_actionMenu.ActionMenu, {
    actionStates: menuActionStates,
    isOpen: props.isMenuOpen,
    onOpen: props.onMenuOpen,
    onClose: props.onMenuClose,
    disabled: props.disabled
  }));
}

function DocumentStatusBarActions(props) {
  var editState = (0, _reactHooks.useEditState)(props.id, props.type);
  var connectionState = (0, _reactHooks.useConnectionState)(props.id, props.type);

  var _useState3 = (0, _react.useState)(false),
      _useState4 = _slicedToArray(_useState3, 2),
      isMenuOpen = _useState4[0],
      setMenuOpen = _useState4[1];

  var actions = editState ? (0, _resolver.default)(editState) : null;
  var handleMenuOpen = (0, _react.useCallback)(() => setMenuOpen(true), []);
  var handleMenuClose = (0, _react.useCallback)(() => setMenuOpen(false), []);
  var handleActionComplete = (0, _react.useCallback)(() => setMenuOpen(false), []);

  if (!actions) {
    return null;
  }

  return /*#__PURE__*/_react.default.createElement(_utils.RenderActionCollectionState, {
    component: DocumentStatusBarActionsInner,
    isMenuOpen: isMenuOpen,
    showMenu: actions.length > 1,
    onMenuOpen: handleMenuOpen,
    onMenuClose: handleMenuClose,
    onActionComplete: handleActionComplete,
    actions: actions,
    actionProps: editState,
    disabled: connectionState !== 'connected'
  });
}

var historyActions = [_HistoryRestoreAction.HistoryRestoreAction];

function HistoryStatusBarActions(props) {
  var editState = (0, _reactHooks.useEditState)(props.id, props.type);
  var connectionState = (0, _reactHooks.useConnectionState)(props.id, props.type);

  if (!editState) {
    return null;
  }

  var disabled = (editState.draft || editState.published || {})._rev === props.revision;

  var actionProps = _objectSpread(_objectSpread({}, editState), {}, {
    revision: props.revision
  });

  return /*#__PURE__*/_react.default.createElement(_utils.RenderActionCollectionState, {
    component: DocumentStatusBarActionsInner,
    actions: historyActions,
    actionProps: actionProps,
    disabled: connectionState !== 'connected' || Boolean(disabled)
  });
}
