// @sanity/form-builder/lib/sanity/DefaultAssetSource/AssetMenu.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AssetMenu = AssetMenu;

var _react = _interopRequireDefault(require("react"));

var _icons = require("@sanity/icons");

var _ui = require("@sanity/ui");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function AssetMenu(_ref) {
  var isSelected = _ref.isSelected,
      onAction = _ref.onAction;
  return /*#__PURE__*/_react.default.createElement(_ui.MenuButton, {
    button: /*#__PURE__*/_react.default.createElement(_ui.Button, {
      mode: "ghost",
      icon: _icons.EllipsisVerticalIcon
    }),
    id: "asset-menu",
    portal: true,
    menu: /*#__PURE__*/_react.default.createElement(_ui.Menu, null, /*#__PURE__*/_react.default.createElement(_ui.MenuItem, {
      text: "Find usages",
      icon: _icons.LinkIcon,
      onClick: () => {
        onAction({
          type: 'showUsage'
        });
      }
    }), /*#__PURE__*/_react.default.createElement(_ui.MenuItem, {
      text: "Delete",
      icon: _icons.TrashIcon,
      tone: isSelected ? undefined : 'critical',
      disabled: isSelected,
      title: isSelected ? 'Cannot delete current image' : undefined,
      onClick: () => {
        onAction({
          type: 'delete'
        });
      }
    })),
    placement: "right"
  });
}
