// @sanity/form-builder/lib/sanity/DefaultAssetSource/DeleteAssetErrorDialog.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DeleteAssetErrorDialog = DeleteAssetErrorDialog;

var _react = _interopRequireDefault(require("react"));

var _icons = require("@sanity/icons");

var _ui = require("@sanity/ui");

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _Preview = _interopRequireDefault(require("../../Preview"));

var _legacyParts = require("../../legacyParts");

var _SpinnerWithText = require("../../components/SpinnerWithText");

var _IntentLink = require("../../transitional/IntentLink");

var _templateObject;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var DocumentLink = (0, _styledComponents.default)(_IntentLink.IntentLink)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  color: inherit;\n  text-decoration: inherit;\n  display: flex;\n  align-items: center;\n  justify-content: space-between;\n"])));

function DeleteAssetErrorDialog(_ref) {
  var asset = _ref.asset,
      onClose = _ref.onClose,
      error = _ref.error;
  return /*#__PURE__*/_react.default.createElement(_ui.Dialog, {
    id: "asset-dialog",
    header: "Image could not be deleted",
    width: 2,
    onClose: onClose
  }, /*#__PURE__*/_react.default.createElement(_ui.Card, {
    padding: 4,
    tone: "critical"
  }, /*#__PURE__*/_react.default.createElement(_ui.Grid, {
    gap: 2,
    style: {
      gridTemplateColumns: 'max-content 1fr'
    }
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    padding: 2
  }, /*#__PURE__*/_react.default.createElement("img", {
    src: "".concat(asset.url, "?w=200"),
    style: {
      maxWidth: '200px'
    },
    alt: "The image that is in use by the listed documents"
  })), /*#__PURE__*/_react.default.createElement(_ui.Box, {
    paddingX: 4
  }, /*#__PURE__*/_react.default.createElement(_legacyParts.WithReferringDocuments, {
    id: asset._id
  }, _ref2 => {
    var isLoading = _ref2.isLoading,
        referringDocuments = _ref2.referringDocuments;
    var drafts = referringDocuments.reduce((acc, doc) => doc._id.startsWith('drafts.') ? acc.concat(doc._id.slice(7)) : acc, []);
    var filteredDocuments = referringDocuments.filter(doc => !drafts.includes(doc._id));

    if (isLoading) {
      return /*#__PURE__*/_react.default.createElement(_SpinnerWithText.SpinnerWithText, {
        text: "Loading..."
      });
    }

    if (filteredDocuments.length === 0) {
      return /*#__PURE__*/_react.default.createElement(_ui.Text, {
        as: "p"
      }, "Could not delete image: ", error.message);
    }

    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_ui.Heading, {
      as: "h4",
      size: 1
    }, filteredDocuments.length > 1 ? /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, filteredDocuments.length, " documents are using this image") : /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "One document is using this image")), /*#__PURE__*/_react.default.createElement(_ui.Box, {
      marginY: 4
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      as: "p"
    }, "Open the document", referringDocuments.length > 1 ? 's' : '', " and remove or replace the image before deleting it.")), /*#__PURE__*/_react.default.createElement(_ui.Grid, {
      gap: 2
    }, filteredDocuments.map(doc => /*#__PURE__*/_react.default.createElement(_ui.Box, {
      key: doc._id
    }, /*#__PURE__*/_react.default.createElement(DocumentLink, {
      intent: "edit",
      params: {
        id: doc._id
      },
      key: doc._id
    }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
      flex: 1
    }, /*#__PURE__*/_react.default.createElement(_Preview.default, {
      value: doc,
      type: _legacyParts.schema.get(doc._type)
    })), /*#__PURE__*/_react.default.createElement(_ui.Box, null, /*#__PURE__*/_react.default.createElement(_ui.Text, null, /*#__PURE__*/_react.default.createElement(_icons.LinkIcon, null), " Open")))))));
  })))));
}
